### DISCUSSION TITLE
Multiple language dialects

### DISCUSSION THREAD

**Discussion by The-Futurist** [antlr/antlr4#4023] [Ideas]
I was pleased to read about this in Parr's book, a good book it is too!

The examples are enough to give me a basis for supporting multiple dialects, I tried this and it works fine:

```
GOTO:       {langcode=="en"}? 'goto' | {langcode=="fr"}? 'gggg' ;
PROCEDURE:  ('procedure' | 'proc') ;
```

Here if the source uses English "goto" its recognized when we set langcode to "en" at runtime, and setting it to "fr" works too.

But is there a way to group these by "langcode"
```
{langcode="en"}
 GOTO:   'goto';
 PROC:   'proc';

{langcode="fr"}
 GOTO:   'gggg';
 PROC:   'pppp';
```

Is there any such notation possible?

**Comment by kaby76** [comment]
The [Antlr4 grammar](https://github.com/antlr/antlr4/blob/76fa05c21b12b96a6c12a0a82e611ed9d87d5af4/tool/src/org/antlr/v4/parse/ANTLRParser.g) doesn't recognize your proposed syntax. The main problem with this syntax is that the action code itself is target specific. In fact, all action code is in a specific target language, even though it can be made to "look" target agnostic by pushing all the code into a method call.. This syntactic extension could be handled by a tree write system, but there isn't one available yet.

**Reply by The-Futurist** [reply]
> The [Antlr4 grammar](https://github.com/antlr/antlr4/blob/76fa05c21b12b96a6c12a0a82e611ed9d87d5af4/tool/src/org/antlr/v4/parse/ANTLRParser.g) doesn't recognize your proposed syntax. The main problem with this syntax is that the action code itself is target specific. In fact, all action code is in a specific target language, even though it can be made to "look" target agnostic by pushing all the code into a method call.. This syntactic extension could be handled by a tree write system, but there isn't one available yet.

Yes, I understand that, I was trying to convey my goal not suggesting that actual syntax. I want to have a set of English keywords, then a set of French, German etc, be able to add more over time.  I'm trying to find the best way to do that in Antlr. Is there a way for example to embed code that could return the token ID from a dictionary indexed by language code or something?

**Reply by kaby76** [reply]
In short, there is no syntax for this. You either need to write an extension of an Antlr grammar and rewrite it into an Antlr grammar, or you implement the code directly with actions.
